src/benchmark.zig

```diff
@@ -234,48 +245,65 @@ const TimedQueue = struct {
-        switch (operation) {
-            .create_accounts => {},
-            .create_transfers => {
-                if (latency > self.max_transfers_latency) {
-                    self.max_transfers_latency = latency;
-                }
-            },
-            .commit_transfers => {
-                if (latency > self.max_commits_latency) {
-                    self.max_commits_latency = latency;
-                }
-            },
-            else => unreachable,
-        }
+            const self: *TimedQueue = @intToPtr(*TimedQueue, @intCast(usize, user_data));
+            const completed_batch: ?Batch = self.batches.pop();
+            assert(completed_batch != null);
+            assert(completed_batch.?.operation == operation);
 
-        var batch: ?Batch = self.batches.peek();
-        if (batch) |*next_batch| {
-            self.batch_start = std.time.milliTimestamp();
-            self.client.request(
-                @intCast(u128, @ptrToInt(self)),
-                TimedQueue.lap,
-                next_batch.operation,
-                next_batch.data,
-            );
-        } else {
-            log.debug("stopping timer...", .{});
-            self.end = now;
+            log.debug("completed batch operation={} start={}", .{
+                completed_batch.?.operation,
+                self.batch_start,
+            });
+            const latency = now - self.batch_start.?;
+            switch (operation) {
+                .create_accounts => {},
+                .create_transfers => {
+                    if (latency > self.max_transfers_latency) {
+                        self.max_transfers_latency = latency;
+                    }
+                },
+                .commit_transfers => {
+                    if (latency > self.max_commits_latency) {
+                        self.max_commits_latency = latency;
+                    }
+                },
+                else => unreachable,
+            }
+
+            var batch: ?Batch = self.batches.peek();
+            var message = self.client.get_message() orelse {
+                @panic("Client message pool has been exhausted.");
+            };
+            defer self.client.unref(message);
+
+            if (batch) |*next_batch| {
+                std.mem.copy(
+                    u8,
+                    message.buffer[@sizeOf(Header)..],
+                    std.mem.sliceAsBytes(next_batch.data),
+                );
+                self.batch_start = std.time.milliTimestamp();
+                self.client.request(
+                    @intCast(u128, @ptrToInt(self)),
+                    TimedQueue.lap,
+                    next_batch.operation,
+                    message,
+                    next_batch.data.len,
+                );
+            } else {
+                log.debug("stopping timer...", .{});
+                self.end = now;
+            }
+        } else |_| {
+            @panic("Client returned error during benchmarking.");
```

Comment:
Would be great to have a `log.emerg()` here using `@errName` to log the actual error just before panicking.

Reply:
To remove the indentation, we could also handle this error branch right up top:

```
const value = results catch {
     log.emerg()
     @panic()
}
```

Reply:
@jorangreef thanks, makes it so much cleaner :+1: